### ISSUE TITLE
Any reason why CronetBuilder doesn't allow custom nameResolverProvider?

### ISSUE THREAD

**Issue by jchau207** [grpc/grpc-java#11067]
The CronetChannelBuilder only uses DirectAddressNameResolver and doesn't allow any new NameResolverProvider to be registered to it.

```
    reg.register(new DirectAddressNameResolverProvider(directServerAddress,
        authority));
    this.nameResolverRegistry = reg;

...

Preconditions.checkState(directServerAddress == null,
        "directServerAddress is set (%s), which forbids the use of NameResolverFactory",
        directServerAddress);

```

**Comment by ejona86** [comment]
Cronet uses the DNS client built into Cronet, so it disables the normal name resolution. Are you needing custom name resolution for Cronet?

**Comment by jchau207** [comment]
> Cronet uses the DNS client built into Cronet, so it disables the normal name resolution. Are you needing custom name resolution for Cronet?

Yes, we have a custom name resolver that uses cloudflare dns service instead Google's. It is more performant from our testing.

**Comment by ejona86** [comment]
The custom name resolver is faster than cronet's, or 1.1.1.1 is faster than 8.8.8.8? I'd expect cronet to be using the device's configured dns server.

Is your app gRPC-only, or does it do any plain HTTP? In mixed apps, there is value in using plain cronet, as you only pay DNS lookup and connection establishment costs once.

Cronet handles TLS and is only given a URL to contact, so it isn't really possible to specify IP addresses from within gRPC. You might take a look at [CronetEngine.Builder](https://developer.android.com/develop/connectivity/cronet/reference/org/chromium/net/CronetEngine.Builder.html) for some of the options it offers, but I'm not seeing anything off-the-top that looks powerful enough for what you need.

**Comment by jchau207** [comment]
We see 1.1.1.1 being faster than 8.8.8.8. 

In terms of communicating with our backend, our app is grpc-only.

I noticed there is a DNS option in the documentation, but I can't access it with the coronet engine builder from the grpc package `io.grpc:grpc-cronet`. Is the coronet builder in io.grpc, a customized builder?

**Comment by ejona86** [comment]
> but I can't access it with the coronet engine builder from the grpc package io.grpc:grpc-cronet

What documentation? You pass [a CronetEngine to the grpc builder](https://github.com/grpc/grpc-java/blob/167a2031e29d21710e8790f72a00589fc6333d59/cronet/src/main/java/io/grpc/cronet/CronetChannelBuilder.java#L68), so you are in full control of Cronet's configuration to whatever extent that is possible.

**Comment by ejona86** [comment]
Seems like we've answered this as much as we can. All I/O within the scope of Cronet itself. If not and we missed something, comment, and it can be reopened.
